nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py

```diff
@@ -1162,11 +1159,12 @@ def on_test_epoch_end(self):
         logging.info(f'test_loss: {averaged_loss[0]}')
         self.test_step_outputs.clear()  # free memory
 
-    def loss_func(self, loss_mask, num_valid_tokens_in_ub, output_tensor):
+    def loss_func(self, loss_mask, num_valid_tokens_per_sample, output_tensor):
         losses = output_tensor.float()
-        loss_mask = loss_mask.view(-1).float()
-        # TODO: add nemo version here
-        loss = torch.sum(losses.view(-1) * loss_mask) / num_valid_tokens_in_ub  # sequence level nll
+        loss_mask = loss_mask.float()
+        # Compute (per-sample) sequence-level NLL. Fully masked samples have zero loss.
+        loss = torch.sum(losses * loss_mask, dim=1) / num_valid_tokens_per_sample.clamp(min=1)
```

Comment:
For pretraining, we don't need per-sample average but per valid token average. Are we going to support these two types of losses?

Reply:
I'm not planning to implement per-valid-token average in this PR, though I agree it could be worth adding in the future. I'm not sure how pre-training uses the loss mask though, it seems to me it should be possible to concatenate data such that there is (almost) no data being masked out (but I may be wrong).

To be sure it's clear: note that the previous code was doing per-sample average with `micro_batch_size=1` (this PR isn't changing this).

Reply:
hmm, before it calculated the per token average per micro-batch. I think it is never per-token loss for the whole batch. It is a bug for pretraining that we need to fix. is there a separate PR to address it?